### PR TITLE
Link UI: Design Iteration

### PR DESCRIPTION
**Pull request by shaunandrews**
The current Link UI (`link-control`) looks a little out of place with the latest block UI (See #26351). Here's how the current Link UI looks and works:

Current Link UI

This PR reduces much of the footprint of the interface, and aims to streamline the flow. Here's how this PR looks now:

Link UI with Settings

Here's some of the changes I've made:

- Reduced much of the whitespace avoiding unnecessarily obscuring the canvas.
- Switched the popover to use the `isAlternate` property which gives the UI dark borders and removes the drop-shadow.
- Added an explicit "Visit" button, rather than linking the URL text.
- The "Edit" action is now a primary button.
- I've removed the confusing "URL" and "Press ENTER" option.
- The "Open in new tab" option now only appears when creating or editing a link.
- Updated the design of the "Create new" button that appears when using the Navigation Link block.

### CODE REVIEW

**Comment by Soean** [comment]
Whats the reason for removing the "open in a new tab" options?

**Comment by shaunandrews** [comment]
> Whats the reason for removing the "open in a new tab" options?

It felt weird being there all the time. I definitely want to bring it back, just haven't gotten to it yet.

**Comment by jasmussen** [comment]
This is a wild cleanup, nice! And the use of the block-UI chrome for the link dialog feels perfect here, it makes me think we should use the same material for this one:

Screenshot 2021-02-05 at 09 10 50

A few notes from testing. In this testing, the word "wondered" is highlighted. Before:

Screenshot 2021-02-05 at 09 13 32

Screenshot 2021-02-05 at 09 13 39

After:

Screenshot 2021-02-05 at 09 12 38

Screenshot 2021-02-05 at 09 12 51

Notes:

- The word highlight doesn't work, this is a separate issue (two selections, one in the canvas, one in the input field). I think it's something we can fix at some point in the future, but it's non-trivial.
- Nevertheless, the new link dialog is not centered below the word to be linked, which makes that disconnect a bit more jarring.
- The ability to just create URL that links to "hello" was present before. I guess it has value if you want to link to `about:blank`, but it appears to be gone now. Intentional or oversight?

Like Soean I think we should keep the ability to open externally, because we have no other good place to put that control. But your reduced design, I think, can have the cake and eat it too, like so: when you first open the dialog and there's just an input field, it's just that, an input field:

Screenshot 2021-02-05 at 09 12 38

Then after you start typing, the option would pop out at the bottom, here:

Screenshot 2021-02-05 at 09 12 51

I think it would fit right into the flow.

Finally, your PRs of late have been _amazing_ 👌  — one small tip, several smaller cleanup PRs might be less overwhelming to tackle from a code standpoint, so we can get the goodness faster! Nice work Shaun.

**Comment by paaljoachim** [comment]
The issue was discussed during a Design Triage. Link to Slack: https://wordpress.slack.com/archives/C02S78ZAL/p1612888787368200

**Comment by shaunandrews** [comment]
I've pushed a bunch of changes to this PR and updated the description/GIFs above. The big changes are bringing back the "Open in new tab" option and adjustments to creating a new draft post/page within the Navigation Link block.

image

> the new link dialog is not centered below the word to be linked, which makes that disconnect a bit more jarring.

I've reverted the change and the dialog is now centered again.

> The ability to just create URL that links to "hello" was present before. I guess it has value if you want to link to about:blank, but it appears to be gone now. Intentional or oversight?

Its still there. Are you unable to do so? Just type _anything_ in the url input and hit enter or click the submit button.

**Comment by joyously** [comment]
> I've reverted the change and the dialog is now centered again.

Did you test it trying to link the last word of the line of text? Where is the dialog placed then?

Do you have to select text first, to create a link?

**Comment by shaunandrews** [comment]
> Did you test it trying to link the last word of the line of text? Where is the dialog placed then?

Yes. It's centered below the word.

> Do you have to select text first, to create a link?

No. If no text is selected the URL will be inserted as linked text where the caret currently resides.

**Comment by kellychoffman** [comment]
Nice clean up. 

Something is off about this step: 
Screen Shot 2021-02-17 at 6 31 51 PM

The URL feels almost wasted in that you can't click on it to edit _or_ to view the link. I think you should be able to do one or the other. I want to just skip right past this whole screen and land right in the edit view, with the addition of a `view` button.

**Comment by joyously** [comment]
> The URL feels almost wasted in that you can't click on it to edit _or_ to view the link. 

That makes no sense to me. "Edit" means edit the URL, not the page it points to.
I'm wondering what is put in for a scheme if none is typed. Or is it only whatever you type goes into the href?

**Comment by shaunandrews** [comment]
> The URL feels almost superfluous in that you can't click on it to edit or to view the link.

I had the URL linked (instead of the Visit button) and I had it as a button (instead of the Edit button), but decided that it was more clear to have explicit Edit and Visit buttons. The URL is purely informational, and I guess could be useful to copy the text if you needed.

> I'm wondering what is put in for a scheme if none is typed. Or is it only whatever you type goes into the href?

Do you mean the protocol? As far as I can tell, `http://` is prepended by default. If you type `https://` (or something else) then it will use that instead. This PR doesn't affect this behavior; This is how it works in `master`.

**Comment by jameskoster** [comment]
Couple of minor comments:

Should the whole link change color on hover here?

hover

---

Screenshot 2021-02-18 at 10 36 21

This icon has always felt a little ambiguous to me. Now that we have the "Edit" and "Visit" buttons in the other state, could this also be a "Done" button or similar?

Speaking of the "Edit" and "Visit" buttons, I noticed that the order of these buttons is inconsistent with other situations in which primary / regular buttons appear adjacently (Top bar, modals). Is there value in being consistent there? 🤔 

---

Small z-index issue occuring here:

Screenshot 2021-02-18 at 10 38 11

---

I think I saw this mentioned elsewhere, but in the old design the selected text remains subtly highlighted when adding a link:

Screenshot 2021-02-18 at 10 44 05

This doesn't seem to have carried over to the new design.

---

Excellent work here, I'm excited for this update :D

**Comment by shaunandrews** [comment]
> I think I saw this mentioned elsewhere, but in the old design the selected text remains subtly highlighted when adding a link... This doesn't seem to have carried over to the new design.

This is currently broken (in `master`) in the post editor, but as you've discovered it works in the site editor. This is a separate issue and unrelated to this PR.

I'll look at the other issues and see what I can do.

**Comment by shaunandrews** [comment]
I've pushed some updates to this PR. Here's how it looks now:

Link UI Again

- I removed the "Visit" button and instead made the URL text a button that opens the link in a new tab.
- The submit icon button has been replaced with a primary "Done" button.
- Hovering a result now correctly changes the title's text color.

--

The z-index issue @jameskoster mentions exists in `master`. I've created a new issue to track and help resolve it: https://github.com/WordPress/gutenberg/issues/29158

**Comment by paaljoachim** [comment]
Comparing versions.

WP 5.6.1 version. (GB not activated)
Screen Shot 2021-02-19 at 21 31 38

PR version.
Screen Shot 2021-02-19 at 21 32 44

The version in this PR is a lot tighter and takes less space.
A Done button is much easier to understand then the return/enter symbol. It makes it easier for new users to remember to click the Done button after they have finished adding the link.

WP 5.6.1 version. (GB not activated)
Screen Shot 2021-02-19 at 21 36 25

PR version.
Screen Shot 2021-02-19 at 21 37 58

From the PR version I am wondering if we should add back in the open in new tab icon seen in the WP 5.6.1 version next to the link.

Bottom line is that the newest iteration feels really well done.

**Comment by mtias** [comment]
I think the biggest tension here is how much it should resemble the toolbar UI vs other UI. The hierarchy offered by the toolbar having a stronger contrast can be useful.

image

For example, when compared to the quick inserter:

image

The Link popover used in Buttons seems to need some attention:

image

This state also looks a bit weird:

image

Should this also be updated at the same time?

image

---

By the way, it's really unfortunate that once you open the link url popover the highlight is lost on the original text, so you have no idea what the link is applying to. cc @ellatrix

**Comment by shaunandrews** [comment]
> I think the biggest tension here is how much it should resemble the toolbar UI vs other UI. The hierarchy offered by the toolbar having a stronger contrast can be useful.

I switched to the darker borders (`isAlternate`) since the popover is triggered _from_ the toolbar, so it felt appropriate for it to share the dark border of the toolbar.

--

I've updated the Buttons block to also use the `isAlternate` styling for the popover.

--

The Social Link block doesn't make use of the `LinkControl` component. The Image block also has its own customized version of a link popover. I expect we can go around updating all the other link-like UI's to use `LinkControl`, but that seems unrelated to this PR.

> This state also looks a bit weird:

I was trying to mimic the search input from the inserters. If we don't think it works, I can update it to look more like a standard input with a dark border instead of a background color.

--

> By the way, it's really unfortunate that once you open the link url popover the highlight is lost on the original text, so you have no idea what the link is applying to.

This is briefly discussed above. I'm not sure we ever had a way to keep the selected text selected, but its worth noting that this _does_ work in the site editor. I believe this is because the site editor uses an `iframe`, which allows us to keep the selection within the canvas _and_ have focus within the popover at the same time. When/If the `iframe` comes to the post editor, this is one of the big improvements we'll see.

**Comment by getdave** [comment]
Closing as the UI has evolved significantly since this.
